### DISCUSSION TITLE
[RTKQ] Inject Multiple Endpoints

### DISCUSSION THREAD

**Discussion by jd-carroll** [reduxjs/redux-toolkit#4576] [Q&A]
Our application is entirely built on AWS Lambda functions and cold starts suck. To help address the cold start problem we've added the ability to "warm" the lambda with a direct API call. So if the normal endpoint is `POST /my/cool-api` the warming endpoint is `POST /my/cool-api -H { x-warming: true }`. The application will generally know when it will potentially need the API so we can issue these warming requests when needed.

NOTE: Yes I know about reserved concurrency and that there are lots of details / nuances to whether or not the warming API call will in fact prevent a cold start. But we are working without a budget here so trying to be efficient 😁

**Question: Is it possible to change how endpoints are injected into the API such that each "definition" results in two injected endpoints (standard + warmer)?**

I'm fairly certain it is, I am just a little lost on the actual implementation. I know I would need to write a custom module, which I think would look something like:

```typescript
export const warmingModule = (): Module<WarmingModule> => ({
  name: warmingModuleName,
  init(api, options, context) {
    // initialize stuff here if you need to

    return {
      injectEndpoint(endpoint, definition) {
        const anyApi = api as any as Api<
          any,
          Record<string, any>,
          string,
          string,
          CustomModule | CoreModule
        >
        // The standard API is already injected so I should only need to add the warmer API
        const warmingEndpoint = `${endpoint}Warmer`;
        anyApi.endpoints[warmingEndpoint] = createNewQueryFromDefinition(definition);
        anyApi.endpoints[warmingEndpoint].warmingFlagOurCustomFetchQueryUses = true;
      },
    }
  },
})
```

But I'm not sure if that is the correct approach as I can't get the types to work (which is probably a separate question). So, ignoring the types for now, **what would be the correct approach to inject an entirely new endpoint?**

**Comment by phryneas** [comment]
Phew... you're the first person to ever ask _that_, so we don't have a recipe for that.

Could you maybe do something like

```ts
export const warmingModule = (): Module<WarmingModule> => ({
  name: warmingModuleName,
  init(api, options, context) {
    // initialize stuff here if you need to

    return {
      injectEndpoint(endpoint, definition) {
        const anyApi = api as any as Api<
          any,
          Record<string, any>,
          string,
          string,
          CustomModule | CoreModule
        >;
        // prevent recursion
        if (endpoint.endsWith("Warmer")) return;
        anyApi.injectEndpoints(/* the warmer endpoint */);
      },
    };
  },
});
```
